### ISSUE TITLE
Unpacking solution fails with NullRefException when unpacking a .msapp

### ISSUE THREAD

**Issue by concoursdude** [microsoft/powerplatform-build-tools#194]
Greetings. 

After working for a couple of weeks now with the folks over on the coe-starter-kit side of the fence, I am reaching out in hopes that someone can assist me in troubleshooting issues with unpacking/packing solutions containing Canvas Apps in my Power Platform ALM pipelines. 

We were hopeful that the September release of the ALM accelerator would fix issues related to packing and unpacking apps as the newest release transitioned from using the PP language tooling to the Power Platform Built Tools tasks, but my failures are persisting. What I'm trying to determine now is if this is an issue with my solution packages (this fails on all solutions in the environment in which I am attempting to commit from), or if this is an issue with the build tasks.

In all scenarios, I receive the error shown in the screenshot below. In the ALM accelerator August hotfix release, committing to source control for these solutions worked, but packing the apps to deploy downstream would result in this failure. In the September release, I am seeing it when unpacking the solutions to my repo during commit.  

Please let me know what other information I can provide in order to resolve this. Thanks.

![image](https://user-images.githubusercontent.com/51216063/189159397-1d010b62-b977-4556-ab23-00164876eb60.png)

**Comment by concoursdude** [comment]
Also wanted to note that I was asked to create a repro of this issue and, using the pipeline defined below, the unpack step is successful for the same solution package. 

![image](https://user-images.githubusercontent.com/51216063/189163777-742b90e7-5568-4bbb-99e9-33653457e3d6.png)

**Comment by concoursdude** [comment]
Also wanted to note that I have created a reproduction of this on a smaller scale using the YAML below and the resulting output is the same error for the unpacking task: 

![image](https://user-images.githubusercontent.com/51216063/189171020-2835603f-28a8-4ede-918f-1c1cb5ff43ba.png)

**Comment by devkeydet** [comment]
Hi @concoursdude - Sorry to hear you are having such difficulties.  There are issues that often surface through the accelerator that other teams own fixing.  This can sometimes be challenging to diagnose and figure out which team owns fixing.  When people have issues with canvas pack/unpack, we typically direct them to this repo:
https://github.com/microsoft/PowerApps-Language-Tooling

That's the repo for the underlying code that performs the unpack/pack of canvas apps.  We are just a wrapper around the devops task.  Similarly, the devops task and pac cli underneath are just a wrapper around the dll built from the repo above.  So, if you can repro it through task, in isolation of the accelerator, my recommendation would be to file an issue using the task repo:
https://github.com/microsoft/powerplatform-build-tools

But bear in mind that they could also conclude this is in the library and ask you to transfer the issue here:
https://github.com/microsoft/PowerApps-Language-Tooling

We recognize it can be a little daunting to have to jump from repo to repo once root cause is identified.  It's hard for us, the accelerator team, to tell from what you've reported whether the root cause is the task or the dll the task uses via pac cli.  But the tasks repo team likely has more access to telemetry than we do to help determine whether it's in the task/cli or the library.

HTH

**Comment by concoursdude** [comment]
> Hi @concoursdude - Sorry to hear you are having such difficulties. There are issues that often surface through the accelerator that other teams own fixing. This can sometimes be challenging to diagnose and figure out which team owns fixing. When people have issues with canvas pack/unpack, we typically direct them to this repo: https://github.com/microsoft/PowerApps-Language-Tooling
> 
> That's the repo for the underlying code that performs the unpack/pack of canvas apps. We are just a wrapper around the devops task. Similarly, the devops task and pac cli underneath are just a wrapper around the dll built from the repo above. So, if you can repro it through task, in isolation of the accelerator, my recommendation would be to file an issue using the task repo: https://github.com/microsoft/powerplatform-build-tools
> 
> But bear in mind that they could also conclude this is in the library and ask you to transfer the issue here: https://github.com/microsoft/PowerApps-Language-Tooling
> 
> We recognize it can be a little daunting to have to jump from repo to repo once root cause is identified. It's hard for us, the accelerator team, to tell from what you've reported whether the root cause is the task or the dll the task uses via pac cli. But the tasks repo team likely has more access to telemetry than we do to help determine whether it's in the task/cli or the library.
> 
> HTH

This is very helpful, @devkeydet, thank you. Since I believe I have repro'd this at the task level as shown above, I assumed posting my issue here was correct (as you pointed out, the /powerplatform-build-tools repo). If I am confused and in the wrong spot please let me know -- I'm still new to using git. Thank you for your prompt reply!

**Comment by concoursdude** [comment]
Hoping I can get some assistance on this this week. @devkeydet is it possible for me to tag someone for help on this? I'm at a roadblock with a customer due to this error.

**Comment by concoursdude** [comment]
> @concoursdude as @devkeydet already mentioned, this is indeed an issue with the PowerApps-Language-Tooling (PALT) library that our pac CLI calls to. The log/screenshot you showed earlier shows in line 23 & 24 that the unpack operation has started to process the `proe_pdfpackagegenerator_*.msapp` and has then run into the NullRefException.
> 
> I will transfer this issue over to our partner team, in the PA-Language-Tooling GH repo mentioned above. This should repro for you locally as well, using the 1.17.6 pac CLI (you can get it from [here](https://www.nuget.org/packages/Microsoft.PowerApps.CLI/1.17.5): download, call `Unblock-File` on the downloaded .nugpkg, then unzip into a local folder, e.g. %USERPROFILE%\Downloads\pacCLI): then run that local pac.exe with: `pac canvas unpack --msapp <path_to_above_msapp> --sources myapp`
> 
> We did ship our August refresh of pac CLI (1.18.4) which does have an updated version of the PALT library, so it's possible your issue was fixed? Worth a try with either downloading our pac CLI 1.18.4 (similar nuget.org link as above). Also, we shipped an updated version of the PP-BT AzDevOps tasks, which also includes the same 1.18.4 version of pac CLI. Try a rerun and verify in the log that your AzDevOps pipeline is now running the PP-BT UnpackSolution `2.0.7`; check the AzDevOps log file that this is indeed the version your pipeline is running (our task extensio should/will update automatically, but there might be some caching delays)
> 
> CC: @tomjschneider

Thanks for the response, @davidjenni. Looks like my pipeline is still running 2.0.5. Not sure what I'd need to do to get that to the newer release. I'll see if this repros locally per your suggestion with both old and new pac CLI releases in the interim, though.

**Comment by davidjenni** [comment]
@concoursdude the update is automagically, since for our routine refreshes, we only bump the patch part of the version. I've not yet fully figured out in what timeframes AzDevOps orgs pick up newly published extensions... We published ours today Thu noon PDT

**Comment by concoursdude** [comment]
Yeah, still seeing the old versions. I'll try another run in the morning. 
![image](https://user-images.githubusercontent.com/51216063/190526736-0ab27adb-146e-4666-b071-2a2e02925986.png)

**Comment by concoursdude** [comment]
Still seeing 2.0.5 this morning. Looking at the install-pac YAML in the Pipeline templates, I see mention of adding new versions of the pac CLI to the versionDictionary, but I'm unsure if this is just a note saying "this is what happens when we invoke this powershell" or if it's saying the new versions need to be manually added in the YAML where $versionDictionary is defined. I'll try again later today to see if 2.0.7 shows up, still just digging around to see what else I can find in the templates in the meantime. 

![image](https://user-images.githubusercontent.com/51216063/190643773-78d42d91-338c-45e0-b345-73f6b0b87287.png)

**Comment by concoursdude** [comment]
Also, not even sure if install-pac.yml is even used anymore, so the above could be irrelevant.

**Comment by tomjschneider** [comment]
@concoursdude Would you be willing to send me a copy of the .msapp file which is reporting the error in your logs? From the provided screenshot, it looks like it may be called `proe_pdfpackagegenerator_d4497_DocumentUri.mspp`. I can run this through the latest version of our unpack library to see if/what might be causing the issue here.

If so, you can send it over to tschneider at microsoft dot com. If not, we can talk through running this directly yourself and seeing what error information the unpack tooling provides.

**Comment by concoursdude** [comment]
@tomjschneider Happy to send you the msapp file. This is actually happening on all of my solution packages containing canvas apps in this tenant where I'm trying to use the build tasks - so the problem appears to be app-agnostic - but I'll send over that particular msapp file and let you play around.

**Comment by concoursdude** [comment]
Also quick update for @davidjenni: I'm now seeing 2.0.7 and seeing the same error on unpack.

**Comment by tomjschneider** [comment]
@concoursdude Thanks for providing the msapp - this appears to be an issue in our unpacking code. We're looking into this and how to resolve it. We'll also look to see if we can find a workaround for you while we resolve and rollout.

**Comment by concoursdude** [comment]
> @concoursdude Thanks for providing the msapp - this appears to be an issue in our unpacking code. We're looking into this and how to resolve it. We'll also look to see if we can find a workaround for you while we resolve and rollout.

Thanks @tomjschneider. No need for a workaround right now, we're just manually moving the solutions across environments in the interim. Let me know if you need anything else from me.

**Comment by concoursdude** [comment]
@tomjschneider I appreciate you and your team working so quickly to fix this issue. I see that a fix for the exception has been merged to the tooling master here: https://github.com/microsoft/PowerApps-Language-Tooling/pull/441

Will I need to wait for a new release to see those changes reflected? Thanks.

**Comment by tomjschneider** [comment]
@concoursdude That's correct - we need to merge these changes into the PAC CLI and then they will be included in the next release. I plan to try and get those merged in today, or early next week, so the fix should be included in the early October release of the PAC CLI. @davidjenni could speak to a more specific targeted date for that release.

**Comment by tomjschneider** [comment]
@concoursdude To follow up, this fix has been merged into the PAC CLI and is just awaiting the next release.

**Comment by concoursdude** [comment]
@tomjschneider Looks like the newest release of ALM is out. Is it safe to assume the fix is in this version? Thanks in advance. 
https://github.com/microsoft/coe-alm-accelerator-templates/releases/tag/CenterofExcellenceALMAccelerator-October2022

**Comment by devkeydet** [comment]
@concoursdude - The accelerator uses the latest v2 tasks that get published to the marketplace.  So, accelerator using this fix  would not be dependent on an accelerator release.  The accelerator would get any fix as soon as the fix is in the version of pac that ships with the published version of the tasks in the devops marketplace.  I'll have to defer to the build tools team on the answer to your question, which is: "When will the next release of the build tools containing this fix be published to the azure devops marketplace?"  But I wanted to clarify the accelerator nuance here.

**Comment by concoursdude** [comment]
Thank you @devkeydet!

**Comment by davidjenni** [comment]
@concoursdude our monthly refresh cycle starts on the first Monday of a month, so our Sept refresh that has @tomjschneider's fix is in that payload. Our current ETA for that pac CLI update 1.19.x, and thus the updates to the PP-BT tasks, is early to mid next week (contingent of us not discovering ship blocker bugs as we do our final validations)

**Comment by davidjenni** [comment]
Fixed with 2.0.8; note that due to Azure DevOps's caching, it might take several hours until all build agents run with the latest update

https://github.com/microsoft/powerplatform-build-tools/releases/tag/v2.0.8

CC: @concoursdude

**Comment by concoursdude** [comment]
Many thanks to @davidjenni @tomjschneider @devkeydet and your respective teams for all of your work to resolve this. My client will be very pleased.
